src/dune_engine/action.mli

```diff
@@ -151,7 +151,7 @@ module Full : sig
   val make
     :  ?env:Env.t (** default [Env.empty] *)
     -> ?locks:Path.t list (** default [[]] *)
-    -> ?can_go_in_shared_cache:bool (** default [true] *)
+    -> ?can_go_in_shared_cache:bool (** default [false] *)
```

Comment:
Comment is stale?

I'm trying to think of how to make this change less painful for you, and I think that without modifying this default. It's going to be rather painful. The issue is that what we call "user rules" is not just whatever is present in simple_rules, but it includes the ocaml compilation rules, C/C++ rules, cram, ocamlformat, etc. So for all of those rules you want `~can_go_in_shared_cache:DUNE_CACHE_RULES`. The only rules for which you want `can_go_in_shared_cache:true` are those in pkg_rules and fetch_rules.

So I think that you indeed should modify this default. In particular, it should be set to `DUNE_CACHE_RULES`.  This will require setting some sort of mutable reference in `Dune_engine.Clflags.can_go_in_shared_cache_default` and then setting it based on this environment variable.

It's somewhat intrusive, but I don't see how else to do this without pulling one's hair out.

Reply:
I've updated it to read the `ref` which is filled from the environment. As it it shouldn't be reset at runtime, I will remove the `cache-user-rules` field, since it can't be changed in watch mode anyway.